### PR TITLE
SP use same params for Py / C++

### PR DESCRIPTION
**Pull request by breznak**
Set same default parameters in C++ and PY implementations of SP.
SP tie-breaker is way too significant, effects the shape of SDR, fixed here.

TODO: 
- [ ] Needs the C++ part merged, please see the issue.
- [x] remove tieBreaker, as discussed in nupic.core related PR by Scott -- moved to a sepratate issue
- [x] merge cleanup at https://github.com/numenta/nupic/pull/2628 -- not needed, separate issue

Fixes: #2616

### CODE REVIEW

**Comment by breznak** [comment]
I'm having trouble with the global inhibition, in both branches for Py &CPP it should be set correctly to `Global`, but a test still keeps failing: 

```
======================================================================
FAIL: testSameInitParams (__main__.SpatialPoolerCompatabilityTest)
Py and C++ SP implementations must use the same init parameters
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/nupic/research/spatial_pooler_compatability_test.py", line 453, in testSameInitParams
    self.compare(pySp, cppSp)
  File "tests/unit/nupic/research/spatial_pooler_compatability_test.py", line 70, in compare
    cppSp.getGlobalInhibition())
AssertionError: True != False within 7 places
```

File: src/nupic/research/spatial_pooler.py [hunk 1]
```diff
@@ -325,8 +332,9 @@ def __init__(self,
 
     # Initialize a tiny random tie breaker. This is used to determine winning
     # columns where the overlaps are identical.
-    self._tieBreaker = 0.01*numpy.array([self._random.getReal64() for i in
```
Comment (breznak) [inline src/nupic/research/spatial_pooler.py]: I believe the current (0.01xRandom) tie-breaker value is too significant and shapes the SDR, while it should be left to the column inhibition process and this array would really only break ties - a value around the permanence's precision would suffice. What do you think?

**Comment by oxtopus** [comment]
This fails for me, too, when tested against https://github.com/numenta/nupic.core/pull/629

```
========================================================================================================================================================== FAILURES ==========================================================================================================================================================
_____________________________________________________________________________________________________________________________________ SpatialPoolerCompatabilityTest.testSameInitParams ______________________________________________________________________________________________________________________________________
self = <tests.unit.nupic.research.spatial_pooler_compatability_test.SpatialPoolerCompatabilityTest testMethod=testSameInitParams>

    def testSameInitParams(self):
      """Py and C++ SP implementations must use the same init parameters"""
      pySp = PySpatialPooler(
          inputDimensions=[121, 1], columnDimensions=[30, 30])
      cppSp = CPPSpatialPooler(
          inputDimensions=[121, 1], columnDimensions=[30, 30])
    
>     self.compare(pySp, cppSp)

tests/unit/nupic/research/spatial_pooler_compatability_test.py:453: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.unit.nupic.research.spatial_pooler_compatability_test.SpatialPoolerCompatabilityTest testMethod=testSameInitParams>, pySp = <nupic.research.spatial_pooler.SpatialPooler object at 0x1102049d0>
cppSp = <nupic.bindings.algorithms.SpatialPooler; proxy of <Swig Object of type 'nupic::algorithms::spatial_pooler::SpatialPooler *' at 0x1101faab0> >

    def compare(self, pySp, cppSp):
      self.assertAlmostEqual(pySp.getPrecision(),
                             cppSp.getPrecision())
      self.assertAlmostEqual(pySp.getNumColumns(),
                             cppSp.getNumColumns())
      self.assertAlmostEqual(pySp.getNumInputs(),
                             cppSp.getNumInputs())
      self.assertAlmostEqual(pySp.getPotentialRadius(),
                             cppSp.getPotentialRadius())
      self.assertAlmostEqual(pySp.getPotentialPct(),
                             cppSp.getPotentialPct())
      self.assertAlmostEqual(pySp.getGlobalInhibition(),
>                            cppSp.getGlobalInhibition())
E     AssertionError: True != False within 7 places

tests/unit/nupic/research/spatial_pooler_compatability_test.py:70: AssertionError
_____________________________________________________________________________________________________________________________________________ SpatialPoolerTest.testExactOutput ______________________________________________________________________________________________________________________________________________
self = <tests.unit.nupic.research.spatial_pooler_unit_test.SpatialPoolerTest testMethod=testExactOutput>

    def testExactOutput(self):
      """
        Given a specific input and initialization params the SP should return this
        exact output.
    
        Previously output varied between platforms (OSX/Linux etc)
        """
    
      expectedOutput = [57, 80, 135, 215, 280, 281, 350, 431, 534, 556, 565, 574,
                        595, 663, 759, 777, 823, 932, 933, 968, 983, 1031, 1126,
                        1184, 1232, 1262, 1420, 1468, 1479, 1516, 1531, 1585,
                        1655, 1672, 1755, 1906, 1927, 1936, 1939, 1940]
    
      sp = SpatialPooler(
        inputDimensions = [1,188],
        columnDimensions = [2048, 1],
        potentialRadius = 94,
        potentialPct = 0.5,
        globalInhibition = 1,
        localAreaDensity = -1.0,
        numActiveColumnsPerInhArea = 40.0,
        stimulusThreshold = 0,
        synPermInactiveDec = 0.01,
        synPermActiveInc = 0.1,
        synPermConnected = 0.1,
        minPctOverlapDutyCycle=0.001,
        minPctActiveDutyCycle=0.001,
        dutyCyclePeriod = 1000,
        maxBoost = 10.0,
        seed = 1956,
        spVerbosity = 0
    
      )
    
    
      inputVector = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0,
                     1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0,
                     1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                     1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
    
      inputArray = numpy.array(inputVector).astype(realDType)
    
      activeArray = numpy.zeros(2048)
    
      sp.compute(inputArray, 1, activeArray)
    
      # Get only the active column indices
      spOutput = [i for i, v in enumerate(activeArray) if v != 0]
>     self.assertEqual(spOutput, expectedOutput)
E     AssertionError: Lists differ: [57, 69, 72, 80, 135, 153, 215... != [57, 80, 135, 215, 280, 281, 3...
E     
E     First differing element 1:
E     69
E     80
E     
E       [57,
E     -  69,
E     -  72,
E        80,
E        135,
E     -  153,
E        215,
E        280,
E        281,
E        350,
E        431,
E        534,
E        556,
E        565,
E        574,
E        595,
E        663,
E        759,
E        777,
E        823,
E        932,
E        933,
E     +  968,
E     +  983,
E        1031,
E        1126,
E        1184,
E        1232,
E        1262,
E     +  1420,
E        1468,
E        1479,
E        1516,
E        1531,
E        1585,
E        1655,
E        1672,
E        1755,
E        1906,
E        1927,
E        1936,
E        1939,
E        1940]

tests/unit/nupic/research/spatial_pooler_unit_test.py:215: AssertionError
------------------------------------------------------------------------------------------------------------------------------------------------------ Captured stdout -------------------------------------------------------------------------------------------------------------------------------------------------------
New seed generated as: 88337 called by run line 322 -> setUp
===================================================================================================================================== 2 failed, 826 passed, 20 skipped in 72.38 seconds ======================================================================================================================================
```

**Comment by rhyolight** [comment]
Marking in progress until CI passes.

**Comment by breznak** [comment]
> Marking in progress until CI passes.

@rhyolight More like ready & blocked, waiting for cpp counterpart to be merged, work on this PR is completed.

**Comment by rhyolight** [comment]
Thanks for the clarification!

**Comment by rhyolight** [comment]
@breznak Blocked by https://github.com/numenta/nupic.core/issues/628 right?

**Comment by breznak** [comment]
@rhyolight yep

**Comment by rhyolight** [comment]
Removed `ready` label until unblocked and CI passes.

**Pull request closed without merging**